Function_04.py

```diff
@@ -0,0 +1,163 @@
+import pandas as pd
+import numpy as np
+from matplotlib import pyplot as plt
+import seaborn as sns
+from sklearn.metrics.pairwise import cosine_similarity
+import operator
+import statistics
+from sklearn.metrics import mean_squared_error 
+
+
+###########
+###########
+## Function for part III. Selecting similar users
+###########
+###########
+
+
+def similar_users(user_id, matrix, k=6):
+    'This function return a list of k user from the matrix which are the most similar to user_id with cosine similarity'
+    
+    # create a df of just the current user
+    user = matrix[matrix.index == user_id]
+    
+    # and a df of all other users
+    other_users = matrix[matrix.index != user_id]
+    
+    # calc cosine similarity between user and each other user
+    similarities = cosine_similarity(user,other_users)[0].tolist()
+    
+    # create list of indices of these users
+    indices = other_users.index.tolist()
+    
+    # create key/values pairs of user index and their similarity
+    index_similarity = dict(zip(indices, similarities))
+    
+    # sort by similarity
+    index_similarity_sorted = sorted(index_similarity.items(), key=operator.itemgetter(1))
+    index_similarity_sorted.reverse()
+    
+    # grab k users off the top
+    top_users_similarities = index_similarity_sorted[:k]
+    users = [u[0] for u in top_users_similarities]
+    
+    return users
+    
+
+###########
+###########
+## Function for part IV. Predictions on item
+###########
+###########
+
+
+def recommend_item_for_all(user_index, similar_user_indices,max_value, matrix,Product_map, known, items=23):
+    'This function allow us to do recommendation either on known item if known is set to True or unknown if it is set to False'
```

Comment:
'This function allows us to do recommendation either on known item if "known" is set to "True" or unknown if it is set to "False"'